translator/internal/internal_to_oc.go

```diff
@@ -0,0 +1,436 @@
+	switch kind {
+	case data.SpanKindCONSUMER:
+		ocKind = tracetranslator.OpenTracingSpanKindConsumer
+	case data.SpanKindPRODUCER:
+		ocKind = tracetranslator.OpenTracingSpanKindProducer
+	case data.SpanKindUNSPECIFIED:
+	case data.SpanKindINTERNAL:
+	case data.SpanKindSERVER: // explicitly handled as SpanKind
+	case data.SpanKindCLIENT: // explicitly handled as SpanKind
+	default:
+
+	}
+
+	if string(ocKind) == "" {
+		// No matching kind attribute value
+		return nil
+	}
+
+	return stringAttributeValue(string(ocKind))
+}
+
+func stringAttributeValue(val string) *octrace.AttributeValue {
+	return &octrace.AttributeValue{
+		Value: &octrace.AttributeValue_StringValue{
+			StringValue: &octrace.TruncatableString{
+				Value: val,
+			},
+		},
+	}
+}
+
+// OTLP follows the W3C format, e.g. "vendorname1=opaqueValue1,vendorname2=opaqueValue2"
+func traceStateToOC(traceState data.TraceState) *octrace.Span_Tracestate {
+	if traceState == "" {
+		return nil
+	}
+
+	// key-value pairs in the "key1=value1" format
+	pairs := strings.Split(string(traceState), ",")
+
+	entries := make([]*octrace.Span_Tracestate_Entry, 0, len(pairs))
+	for _, pair := range pairs {
+		kv := strings.SplitN(pair, "=", 2)
+		if len(kv) == 0 {
+			continue
+		}
+
+		key := kv[0]
+		val := ""
+		if len(kv) >= 2 {
+			val = kv[1]
+		}
+
+		entries = append(entries, &octrace.Span_Tracestate_Entry{
+			Key:   key,
+			Value: val,
+		})
+	}
+
+	return &octrace.Span_Tracestate{
+		Entries: entries,
+	}
+}
+
+func spanKindToOC(kind data.SpanKind) octrace.Span_SpanKind {
+	switch kind {
+	case data.SpanKindSERVER:
+		return octrace.Span_SERVER
+	case data.SpanKindCLIENT:
+		return octrace.Span_CLIENT
+	// NOTE: see `spanKindToOCAttribute` function for custom kinds
+	case data.SpanKindUNSPECIFIED:
+	case data.SpanKindINTERNAL:
+	case data.SpanKindPRODUCER:
+	case data.SpanKindCONSUMER:
+	default:
+	}
+
+	return octrace.Span_SPAN_KIND_UNSPECIFIED
+}
+
+func eventsToOC(events []*data.SpanEvent, droppedCount uint32) *octrace.Span_TimeEvents {
+	if len(events) == 0 && droppedCount == 0 {
+		return nil
+	}
+
+	ocEvents := make([]*octrace.Span_TimeEvent, 0, len(events))
+	for _, event := range events {
+		ocEvents = append(ocEvents, eventToOC(event))
+	}
+
+	return &octrace.Span_TimeEvents{
+		TimeEvent:                 ocEvents,
+		DroppedMessageEventsCount: int32(droppedCount),
+	}
+}
+
+func eventToOC(event *data.SpanEvent) *octrace.Span_TimeEvent {
+	attrs := event.Attributes()
+
+	ocMessageEventType, ocMessageEventTypeAttrFound := attrs[conventions.OCTimeEventMessageEventType]
+	ocMessageEventID, ocMessageEventIDAttrFound := attrs[conventions.OCTimeEventMessageEventID]
+	ocMessageEventUSize, ocMessageEventUSizeAttrFound := attrs[conventions.OCTimeEventMessageEventUSize]
+	ocMessageEventCSize, ocMessageEventCSizeAttrFound := attrs[conventions.OCTimeEventMessageEventCSize]
+
+	// Consider TimeEvent to be of MessageEvent type if all relevant attributes are set
```

Comment:
👍 

Can we also check that no other attributes are set? If we don't then the attributes will be lost. If there are other attributes it is preferable to use Annotation to avoid data loss.